ai_genomics/analysis/gtr_definitions.py

```diff
@@ -76,49 +75,44 @@
     )
 
     #  Intersection of the two sets
-    unified_projects = set(
-        ai_projs_abstr.union(ai_topics) & genom_projs_abstr.union(genomics_topics)
+    unified_projects = list(
+        set(ai_projs_abstr.union(ai_topics) & genom_projs_abstr.union(genomics_topics))
     )
 
-    # Extract projects (NB there are duplicates in the projects list)
-    rel_projects = [proj for proj in projects if proj["id"] in unified_projects]
-
-    ai_genomics_combined = []
-    used_ids = []
-
-    # We use this to catch dupes
-    for _id in set([proj["id"] for proj in rel_projects]):
-        for p in rel_projects:
-            if _id not in used_ids and p["id"] == _id:
-                ai_genomics_combined.append(p)
-                used_ids.append(_id)
+    ai_genomics_combined = (
+        pd.DataFrame.from_dict(projects)
+        .query(f"id in {unified_projects}")
+        .reset_index(drop=True)
+    )
 
     logging.info(f"AI and genomics projects combined: {len(ai_genomics_combined)}")
 
     # Get project examples
-    project_examples = [
-        {
-            "Project title": sampled["title"],
-            "Abstract (truncated)": sampled["abstractText"][:700] + "...",
-        }
-        for sampled in random.sample(ai_genomics_combined, 5)
-    ]
-
-    pd.DataFrame(project_examples).to_markdown(
-        f"{PROJECT_DIR}/outputs/gtr_examples.md", index=False
+    project_examples = ai_genomics_combined.sample(5)[["title", "abstractText"]].rename(
+        columns={"title": "Project title"}
+    )
+    project_examples["Abstract (truncated)"] = (
+        project_examples["abstractText"].str[:700] + "..."
     )
+    project_examples = project_examples[["Project title", "Abstract (truncated)"]]
+    project_examples.to_markdown(f"{PROJECT_DIR}/outputs/gtr_examples.md", index=False)
 
-    logging.info(pd.DataFrame(project_examples))
+    logging.info(project_examples)
 
     # Save project ids
-    with open(GTR_INPUTS_DIR / "gtr_projects.json", "w") as f:
-        json.dump(ai_genomics_combined, f)
+    ai_genomics_combined.to_json(GTR_INPUTS_DIR / "gtr_ai_genomics_projects.json")
 
+    # Get publications from projects
     publications_from_projects = fetch_gtr("gtr_projects-outcomes_publications")
 
-    # Get publications from projects
-    genom_publs = [
-        pap for pap in publications_from_projects if pap["project_id"] in used_ids
-    ]
+    genom_publs = pd.DataFrame(
+        [
+            pap
+            for pap in publications_from_projects
+            if pap["project_id"] in ai_genomics_combined.id.to_list()
```

Comment:
Not that you would get a noticeable slow down in this particular case, but I was curious about the performance of this so tested this contrived example:

```
df = pd.DataFrame({"a": np.random.randint(1, 10_000, 10_000)})
x = np.random.randint(1, 10_000, 1_000)
```

```
# to_list inside list comprehension
y = [i for i in x if i in df["a"].to_list()]

326 ms ± 3.93 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

```
# to_list pre-computed
l = df["a"].to_list()
y = [i for i in x if i in l]

152 ms ± 500 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

```
# no to_list
y = [i for i in x if i in df["a"]]

2.13 ms ± 30.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```